Support for Kubernetes and Longhorn

jameswynn opened the pull request:
I've added support for monitoring Kubernetes resources (cpu and memory) as well as service discovery. Storage metrics depend heavily on the underlying storage mechanism, so I've added support for Longhorn as well.

shamoon commented:
Seems awesome, thanks for all the work!

Now, how the hell can I test this out 😑, I have 0 kubernetes experience (and dont have the time to learn RN)

Hopefully someone else can...

jameswynn commented:
It should be pretty easy to spin up minikube and deploy it there. I'll try to create some simple instructions. Homepage could also use a Helm chart. Perhaps that will be my next PR.

JazzFisch commented on src/pages/api/kubernetes/stats/[...service].js:
```diff
@@ -0,0 +1,79 @@
+    const pods = podsResponse.body.items;
+
+    if (pods.length === 0) {
+      res.status(200).send({
```
Should this be a 404 instead of a 200?

JazzFisch commented on src/pages/api/kubernetes/stats/[...service].js:
```diff
@@ -0,0 +1,79 @@
+      stats,
+    });
+  } catch (e) {
+    console.log("error", e);
```
Can you please use the homepage logger?
```js
import createLogger from "utils/logger";
const logger = createLogger("kubernetesStatsService");
logger.error(e);
```

JazzFisch commented on src/pages/api/kubernetes/status/[...service].js:
```diff
@@ -0,0 +1,42 @@
+    const pods = podsResponse.body.items;
+
+    if (pods.length === 0) {
+      res.status(200).send({
```
Can you return 404 here?

JazzFisch commented on src/pages/api/kubernetes/status/[...service].js:
```diff
@@ -0,0 +1,42 @@
+      status
+    });
+  } catch {
+    res.status(500).send({
```
Can you please use the homepage logger?
```js
import createLogger from "utils/logger";
const logger = createLogger("kubernetesStatusService");
logger.error(e);
```

jameswynn commented:
In order to run Homepage in the cluster with discovery and metrics you will need to define a ClusterRole with at least these permissions:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: homepage
rules:
  - apiGroups:
      - ""
    resources:
      - namespaces
      - pods
      - nodes
    verbs:
      - get
      - list
  - apiGroups:
      - extensions
      - networking.k8s.io
    resources:
      - ingresses
    verbs:
      - get
      - list
  - apiGroups:
      - metrics.k8s.io
    resources:
      - nodes
      - pods
    verbs:
      - get
      - list
```

JazzFisch commented:
> In order to run Homepage in the cluster with discovery and metrics you will need to define a ClusterRole with at least these permissions:

Note to self, ensure this gets added to the docs.

jameswynn commented:
I've created a helm chart which should simplify deployment and testing on kubernetes: https://github.com/jameswynn/helm-charts/tree/main/charts/homepage

jameswynn commented:
What do you think about the approach I've taken with the resources widget? I made resources' backend configurable instead of entirely separating it like glances. Would the other approach be better?

jameswynn commented:
[This article](https://techdozo.dev/getting-started-with-kind-quick-start-a-multi-node-local-kubernetes-cluster/) explains how to quickly create a local Kubernetes cluster using Docker. If you follow these instructions you will have a cluster that can readily deploy the application. Once a cluster is ready, you can deploy Homepage using my helm chart and Helm v3:

```sh
helm repo add jameswynn-charts http://jameswynn.github.io/helm-charts
helm repo update
helm install homepage jameswynn-charts/homepage -f values.yaml
```

With a `values.yaml` file like this:

```yaml
enableRbac: true
config:
  kubernetes:
    mode: cluster
  widgets:
    - resources:
        backend: kubernetes
        expanded: true
        cpu: true
        memory: true
ingress:
  main:
    enabled: true
    labels:
      homepage/enabled: "true"
    annotations:
      homepage/name: "Homepage"
      homepage/description: "A modern, secure, highly customizable application dashboard."
      homepage/group: "My Group"
      homepage/icon: "homepage.png"
    ingressClassName: "nginx"
    hosts:
      - paths:
          - path: /
            pathType: Prefix
```

If you followed that article then Homepage should be available at http://localhost:81/

JazzFisch commented:
@jameswynn there are the following warnings when performing a `pnpm install` with your change.

>  WARN  deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
 WARN  deprecated har-validator@5.1.5: this library is no longer supported
 WARN  deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.

Can you please resolve these warnings and update your PR?  Thanks.

jameswynn commented:
Just pushed more changes and merged in the most recent changes from the main branch. `pnpm install` doesn't show any issues on my machine.

Here is a preview of the current Kubernetes widgets. I've made it configurable to either show only the cluster, or all the individual nodes in the cluster. Showing CPU and memory separately was going to take far too much room so I consolidated them.
![image](https://user-images.githubusercontent.com/1775801/202822908-48682987-b59e-4102-933f-25ea027e14fb.png)

This is based on the following `services.yaml` snippet:

```yaml
- kubernetes:
    cluster:
      show: true
      cpu: true
      memory: true
      showLabel: true
      label: "cluster"
    nodes:
      show: true
      cpu: true
      memory: true
      showLabel: true
```

JazzFisch commented:
I really love the name of the machines in your cluster.  😁

clbx commented:
I'm able to get the image working, however neither service discovery or the Kubernetes resources widget seem to work for me and both dump layout errors into the pod logs. 

Could you post your configmap manifest?

jameswynn commented:
This is an abbreviated version of the config that I am currently using. Obviously replace the URLs and ignore the ${SECRET_INTERNAL_DOMAIN} bit as that is for my Flux CD configuration. I think the combination of "app" and "namespace" to select applications is probably too simplistic.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: homepage
data:
  bookmarks.yaml: |
    ---

  docker.yaml: |
    ---

  kubernetes.yaml: |
    ---
    mode: cluster

  services.yaml: |
    ---
    - Development:
        - Gitea:
            icon: gitea.png
            href: https://gitea.${SECRET_INTERNAL_DOMAIN}
            description: Private Github Alternative
            app: gitea
            namespace: dev
        - Drone CI:
            icon: drone.png
            href: https://drone.${SECRET_INTERNAL_DOMAIN}
            description: CI/CD Platform
            app: drone
            namespace: dev

  settings.yaml: |
    ---
    providers:
      longhorn:
        url: https://longhorn.${SECRET_INTERNAL_DOMAIN}

  widgets.yaml: |
    ---
    - kubernetes:
        cluster:
          show: true
          cpu: true
          memory: true
          showLabel: true
          label: "cluster"
        nodes:
          show: true
          cpu: true
          memory: true
          showLabel: true

    - longhorn:
        expanded: true
        total: true

    - search:
        provider: duckduckgo
        target: _blank

```

omdv commented:
So based on the responses looks like there is no automatic service discovery yet via something like ingress annotations. I will try to look how it is done in e.g. [hajimari](https://hajimari.io/) and maybe take a stab at implementing it.

jameswynn commented:
Actually that is already enabled, as per https://github.com/benphelps/homepage/pull/448#issuecomment-1305720937

clbx commented:
> @clbx did you use the helm chart I mentioned, or deploy it directly? I'll post a sample config map shortly.

I am not using the helm chart, I wrote my manifests before discovering your fork.

jameswynn commented:
I'm updating my fork to line up with the most recent changes, including the new textual statuses.

jameswynn commented:
I've changed the label prefix for autodetection via ingress from "homepage" to "gethomepage.dev" which is more inline with kubernetes practices. So annotations would look like this now:

```yaml
ingress:
  main:
    enabled: true
    labels:
      gethomepage.dev/enabled: "true"
    annotations:
      gethomepage.dev/name: "Homepage"
      gethomepage.dev/description: "A modern, secure, highly customizable application dashboard."
      gethomepage.dev/group: "My Group"
      gethomepage.dev/icon: "homepage.png"
    ingressClassName: "nginx"
    hosts:
      - paths:
          - path: /
            pathType: Prefix
```
I've also added a new option to provide a `podSelector` which takes precedence over the "app" field. This allows for atypical/complex deployments to be better represented. For instance, Longhorn itself won't be fully captured without a podSelector.

Here is an example of a service entry with a podSelector which would show the status of all the components deployed by under the gitea instance. In my case this would include the postgresql and memcached instances.

```yaml
- Development
    - Gitea:
        icon: gitea.png
        href: https://gitea.${SECRET_INTERNAL_DOMAIN}
        description: Private Github Alternative
        app: gitea
        namespace: dev
        podSelector: "app.kubernetes.io/instance=gitea"
```

jameswynn commented:
I'm not sure what needs to be done for this "cannot build offline" issue. Suggestions?

jameswynn commented:
I was able to build it by removing the --offline from the Dockerfile. I'm not sure why we would/wouldn't want that though.

omdv commented:
I am happy to test it on my cluster, @jameswynn do you have a published docker image with all changes?

jameswynn commented:
Here is the image that I use, x86 only: `ghcr.io/jameswynn/homepage:kubernetes`

@clbx Did you configure the cluster role and service account?

clbx commented:
Service auto-discovery, cluster widgets and the podselector all work for me.

<img width="1166" alt="image" src="https://user-images.githubusercontent.com/15034517/206751975-a2975af2-3d3e-4e3c-97aa-4060b32a8ea7.png">

shamoon commented:
Awesome. I know jameswynn you've put a lot of work into this, as I see it the only barrier is maintenance since that may largely fall to you, so I think the decision on this will have to be up to Ben (who as you noticed has been pretty otherwise occupied). Regardless, appreciate the effort 👏

jameswynn commented:
OK. In the meantime I will try to keep my repo, image, and chart up to date with mainline. If we merge the feature, then we should also consider migrating the chart into this repo as well. I can definitely help with setting that up (pretty straightforward) as well as expanding the documentation for this feature.

clbx commented on src/pages/api/widgets/longhorn.js:
```diff
@@ -0,0 +1,82 @@
+    return null;
+  }
+
+  const nodes = json.data.map((node) => {
+    let available = 0;
+    let maximum = 0;
+    let reserved = 0;
+    let scheduled = 0;
+    Object.keys(node.disks).forEach((diskKey) => {
+      const disk = node.disks[diskKey];
+      available += disk.storageAvailable;
+      maximum += disk.storageMaximum;
+      reserved += disk.storageReserved;
+      scheduled += disk.storageScheduled;
+    });
+    return {
+      id: node.id,
+      available,
+      maximum,
+      reserved,
+      scheduled,
+    };
+  });
+  const total = nodes.reduce((summary, node) => ({
+    available: summary.available + node.available,
+    maximum: summary.maximum + node.maximum,
+    reserved: summary.reserved + node.reserved,
+    scheduled: summary.scheduled + node.scheduled,
+  }));
+  total.id = "total";
+  nodes.push(total);
+  return nodes;
+}
```
The longhorn integration does not work for me (just get "API Error")

![Screen Shot 2022-12-09 at 5 47 56 PM](https://user-images.githubusercontent.com/15034517/206808113-9230ddbf-5708-4eeb-b472-3be0caef0de4.png)


I think the issue lies somewhere in here. I believe that having nodes that do not have any disks is causing an error and the pod logs only give:
```
[2022-12-09T22:42:37.206Z] error: undefined
```

The response I get lines up with this, but it might be erroring out on nodes where there are no disks. I'm not a JS dev though, so I could be way off. I'm also hitting the endpoint from inside the cluster, this shouldn't make a difference but important to note.

jameswynn replied:
Good catch! I didn't consider nodes without disks. I'll make some changes.

jameswynn replied:
I've added a conditional here and pushed a new image. Let me know if it fixes the issue.

clbx replied:
Still seeing the same behavior

clbx replied:
```
{
  "data": [
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/apollo-3050-kube2?action=diskUpdate"
      },
      "address": "10.42.2.42",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-14T01:37:46Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-12-12T00:46:04Z",
          "message": "Node apollo-3050-kube2 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-07T22:00:55Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {
        "disk-apollo": {
          "allowScheduling": true,
          "conditions": {
            "Ready": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-07T22:01:00Z",
              "message": "Disk disk-apollo(/mnt/longhorn/) on node apollo-3050-kube2 is ready",
              "reason": "",
              "status": "True",
              "type": "Ready"
            },
            "Schedulable": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-07T22:01:00Z",
              "message": "Disk disk-apollo(/mnt/longhorn/) on node apollo-3050-kube2 is schedulable",
              "reason": "",
              "status": "True",
              "type": "Schedulable"
            }
          },
          "diskUUID": "347246ae-85f1-48f6-a51f-f482b8f13532",
          "evictionRequested": false,
          "path": "/mnt/longhorn/",
          "scheduledReplica": {
            "pvc-456091fa-b653-4750-9d7f-r-72d5d791": 17179869184,
            "pvc-4ea24031-9b54-444a-9625-r-5109e2e6": 68719476736,
            "pvc-526925ff-ea69-4175-b4fb-r-d7f2d9fc": 17179869184,
            "pvc-550b406d-82fa-44ce-b4b1-r-46bc25f8": 4294967296,
            "pvc-5d210118-55aa-4ace-a2b2-r-7d6c8fb1": 34359738368,
            "pvc-753648f8-a74f-496a-8e43-r-f1639c94": 10737418240,
            "pvc-78936a63-a5ab-43f9-8786-r-fad52dd9": 17179869184,
            "pvc-7fea26c6-6f71-4d94-a578-r-df233fab": 17179869184,
            "pvc-98551821-26a2-4c0f-8cc1-r-0bb6398c": 8589934592,
            "pvc-9d3d0e67-2bd5-4b34-9131-r-ddb4d646": 34359738368,
            "pvc-9e5f52ba-3f61-4217-8785-r-7bff517b": 34359738368,
            "pvc-c2087672-d02e-4e6f-907c-r-e1f9fc22": 34359738368,
            "pvc-cd80907a-f9e0-4e37-a168-r-70a342cb": 4294967296,
            "pvc-d21984f2-283d-469f-8903-r-b351f08a": 10737418240,
            "pvc-d375c4a1-dc40-4e8a-8d03-r-20289af5": 8589934592,
            "pvc-dd976c8a-d1b3-4a25-bd86-r-19e86775": 17179869184,
            "pvc-eb71c9a6-304b-4a07-a930-r-4be7a3b4": 34359738368
          },
          "storageAvailable": 879021260800,
          "storageMaximum": 983350071296,
          "storageReserved": 0,
          "storageScheduled": 511101108224,
          "tags": []
        }
      },
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "apollo-3050-kube2",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/apollo-3050-kube2"
      },
      "name": "apollo-3050-kube2",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/artemis-3050-kube3?action=diskUpdate"
      },
      "address": "10.42.3.195",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-14T01:37:44Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-12-12T00:46:04Z",
          "message": "Node artemis-3050-kube3 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-14T01:37:44Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {},
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "artemis-3050-kube3",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/artemis-3050-kube3"
      },
      "name": "artemis-3050-kube3",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/gemini-3050-kube1?action=diskUpdate"
      },
      "address": "10.42.1.232",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-14T01:37:46Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-12-12T00:46:01Z",
          "message": "Node gemini-3050-kube1 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-09T18:36:34Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {},
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "gemini-3050-kube1",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/gemini-3050-kube1"
      },
      "name": "gemini-3050-kube1",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/jamestown-7050-kube4?action=diskUpdate"
      },
      "address": "10.42.5.200",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-26T04:59:48Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-09T19:11:00Z",
          "message": "Node jamestown-7050-kube4 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-09T18:36:34Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {
        "disk-jamestown": {
          "allowScheduling": true,
          "conditions": {
            "Ready": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-09T18:42:05Z",
              "message": "Disk disk-jamestown(/mnt/longhorn/) on node jamestown-7050-kube4 is ready",
              "reason": "",
              "status": "True",
              "type": "Ready"
            },
            "Schedulable": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-09T18:42:05Z",
              "message": "Disk disk-jamestown(/mnt/longhorn/) on node jamestown-7050-kube4 is schedulable",
              "reason": "",
              "status": "True",
              "type": "Schedulable"
            }
          },
          "diskUUID": "80d17318-5684-4f83-879d-b8a5135b4872",
          "evictionRequested": false,
          "path": "/mnt/longhorn/",
          "scheduledReplica": {
            "pvc-456091fa-b653-4750-9d7f-8b80bb589a4f-r-e8a4b0e5": 17179869184,
            "pvc-4ea24031-9b54-444a-9625-10954d7859e8-r-ca7e74b0": 68719476736,
            "pvc-526925ff-ea69-4175-b4fb-4228415bd312-r-ff4f2eb6": 17179869184,
            "pvc-550b406d-82fa-44ce-b4b1-4d5404045316-r-a9bf80cc": 4294967296,
            "pvc-7fea26c6-6f71-4d94-a578-2f13380b4bb9-r-b8bdcdb1": 17179869184,
            "pvc-c2087672-d02e-4e6f-907c-42e031bdd474-r-a937270d": 34359738368,
            "pvc-d375c4a1-dc40-4e8a-8d03-11fcfad885c3-r-194e9666": 8589934592
          },
          "storageAvailable": 960390758400,
          "storageMaximum": 983350071296,
          "storageReserved": 0,
          "storageScheduled": 167503724544,
          "tags": []
        }
      },
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "jamestown-7050-kube4",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/jamestown-7050-kube4"
      },
      "name": "jamestown-7050-kube4",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/mercury-7040-kube0?action=diskUpdate"
      },
      "address": "10.42.0.206",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-14T01:37:44Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-09T19:11:02Z",
          "message": "Node mercury-7040-kube0 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-08T20:25:39Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {
        "disk-mercury": {
          "allowScheduling": true,
          "conditions": {
            "Ready": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-08T20:25:48Z",
              "message": "Disk disk-mercury(/mnt/longhorn/) on node mercury-7040-kube0 is ready",
              "reason": "",
              "status": "True",
              "type": "Ready"
            },
            "Schedulable": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-08T20:25:48Z",
              "message": "Disk disk-mercury(/mnt/longhorn/) on node mercury-7040-kube0 is schedulable",
              "reason": "",
              "status": "True",
              "type": "Schedulable"
            }
          },
          "diskUUID": "eab81173-8191-4afc-af4b-01b9cc51946c",
          "evictionRequested": false,
          "path": "/mnt/longhorn/",
          "scheduledReplica": {
            "jellyfin-r-50871892": 137438953472,
            "pvc-456091fa-b653-4750-9d7f-8b80bb589a4f-r-1a50aab7": 17179869184,
            "pvc-4ea24031-9b54-444a-9625-10954d7859e8-r-d7ee4e4a": 68719476736,
            "pvc-526925ff-ea69-4175-b4fb-4228415bd312-r-b717613f": 17179869184,
            "pvc-550b406d-82fa-44ce-b4b1-4d5404045316-r-048a8be2": 4294967296,
            "pvc-5d210118-55aa-4ace-a2b2-6e1531437062-r-d2f19d2d": 34359738368,
            "pvc-753648f8-a74f-496a-8e43-38737486a37a-r-de5223d6": 10737418240,
            "pvc-78936a63-a5ab-43f9-8786-165311579efc-r-e36938a8": 17179869184,
            "pvc-7fea26c6-6f71-4d94-a578-2f13380b4bb9-r-01f48587": 17179869184,
            "pvc-98551821-26a2-4c0f-8cc1-b5e6d368194b-r-6e1b1db4": 8589934592,
            "pvc-9d3d0e67-2bd5-4b34-9131-4ba878e39725-r-11c69801": 34359738368,
            "pvc-9e5f52ba-3f61-4217-8785-48b97f75ce7d-r-545a52e0": 34359738368,
            "pvc-c2087672-d02e-4e6f-907c-42e031bdd474-r-cbb31415": 34359738368,
            "pvc-cd80907a-f9e0-4e37-a168-bbcf298fb85c-r-349d4eca": 4294967296,
            "pvc-d21984f2-283d-469f-8903-1e243d0879d0-r-cfe0e80d": 10737418240,
            "pvc-d375c4a1-dc40-4e8a-8d03-11fcfad885c3-r-7f2fcca2": 8589934592,
            "pvc-dd976c8a-d1b3-4a25-bd86-e447a13c24af-r-7fe25aa8": 17179869184,
            "pvc-eb71c9a6-304b-4a07-a930-e82be4a085e6-r-5326a234": 34359738368
          },
          "storageAvailable": 879021260800,
          "storageMaximum": 983350071296,
          "storageReserved": 0,
          "storageScheduled": 511101108224,
          "tags": []
        }
      },
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "mercury-7040-kube0",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/mercury-7040-kube0"
      },
      "name": "mercury-7040-kube0",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/perseverance-rpi4-kube5?action=diskUpdate"
      },
      "address": "10.42.4.58",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-23T07:34:10Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-09T19:11:10Z",
          "message": "Node perseverance-rpi4-kube5 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-08-23T07:34:10Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {},
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "perseverance-rpi4-kube5"
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/perseverance-rpi4-kube5"
      },
      "name": "perseverance-rpi4-kube5",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    },
    {
      "actions": {
        "diskUpdate": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/seadragon-dl380-kube8?action=diskUpdate"
      },
      "address": "10.42.6.164",
      "allowScheduling": true,
      "conditions": {
        "MountPropagation": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-08T23:02:43Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "MountPropagation"
        },
        "Ready": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-12-12T00:47:42Z",
          "message": "Node seadragon-dl380-kube8 is ready",
          "reason": "",
          "status": "True",
          "type": "Ready"
        },
        "Schedulable": {
          "lastProbeTime": "",
          "lastTransitionTime": "2022-11-08T23:02:43Z",
          "message": "",
          "reason": "",
          "status": "True",
          "type": "Schedulable"
        }
      },
      "disks": {
        "disk-seadragon": {
          "allowScheduling": true,
          "conditions": {
            "Ready": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-08T23:03:43Z",
              "message": "Disk disk-seadragon(/mnt/longhorn/) on node seadragon-dl380-kube8 is ready",
              "reason": "",
              "status": "True",
              "type": "Ready"
            },
            "Schedulable": {
              "lastProbeTime": "",
              "lastTransitionTime": "2022-11-08T23:03:43Z",
              "message": "Disk disk-seadragon(/mnt/longhorn/) on node seadragon-dl380-kube8 is schedulable",
              "reason": "",
              "status": "True",
              "type": "Schedulable"
            }
          },
          "diskUUID": "ff2cb4db-3b0e-4b01-a491-b9c3bad5755b",
          "evictionRequested": false,
          "path": "/mnt/longhorn/",
          "scheduledReplica": {
            "jellyfin-r-07f1e5aa": 137438953472,
            "pvc-5d210118-55aa-4ace-a2b2-6e1531437062-r-66c9e742": 34359738368,
            "pvc-753648f8-a74f-496a-8e43-38737486a37a-r-053c9e58": 10737418240,
            "pvc-78936a63-a5ab-43f9-8786-165311579efc-r-4a817bc6": 17179869184,
            "pvc-98551821-26a2-4c0f-8cc1-b5e6d368194b-r-8a4e0aee": 8589934592,
            "pvc-9d3d0e67-2bd5-4b34-9131-4ba878e39725-r-efc0392f": 34359738368,
            "pvc-9e5f52ba-3f61-4217-8785-48b97f75ce7d-r-200a91cc": 34359738368,
            "pvc-cd80907a-f9e0-4e37-a168-bbcf298fb85c-r-a742e23e": 4294967296,
            "pvc-d21984f2-283d-469f-8903-1e243d0879d0-r-cc945c54": 10737418240,
            "pvc-dd976c8a-d1b3-4a25-bd86-e447a13c24af-r-15ddca6a": 17179869184,
            "pvc-eb71c9a6-304b-4a07-a930-e82be4a085e6-r-6ffe0491": 34359738368
          },
          "storageAvailable": 1492543078400,
          "storageMaximum": 1574014504960,
          "storageReserved": 0,
          "storageScheduled": 343597383680,
          "tags": []
        }
      },
      "engineManagerCPURequest": 0,
      "evictionRequested": false,
      "id": "seadragon-dl380-kube8",
      "links": {
        "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes/seadragon-dl380-kube8"
      },
      "name": "seadragon-dl380-kube8",
      "region": "",
      "replicaManagerCPURequest": 0,
      "tags": [],
      "type": "node",
      "zone": ""
    }
  ],
  "links": {
    "self": "http://longhorn-frontend.longhorn-system.svc.cluster.local/v1/nodes"
  },
  "resourceType": "node",
  "type": "collection"
}
```

clbx replied:
Sure, send me a DM mine is clay#0124

insuusvenerati commented:
Playing around with this on minikube and i'm getting a 500 in the browser (404 on the server). Any ideas what would cause that? 
My config is just 
```yaml
  widgets:
    - resources:
        # change backend to 'kubernetes' to use Kubernetes integration. Requires RBAC.
        backend: kubernetes
        expanded: true
        cpu: true
        memory: true
    - search:
        provider: duckduckgo
        target: _blank
    - kubernetes:
        cluster:
          show: true
          cpu: true
          memory: true
          showLabel: true
          label: "cluster"
        nodes:
          show: true
          cpu: true
          memory: true
          showLabel: true
  kubernetes:
    # change mode to 'cluster' to use RBAC service account
    mode: cluster
  docker:
  settings:
```

insuusvenerati commented:
> Playing around with this on minikube and i'm getting a 500 in the browser (404 on the server). Any ideas what would cause that? RBAC is enabled and the service account exists. My config is just
> 
> ```yaml
>   widgets:
>     - resources:
>         # change backend to 'kubernetes' to use Kubernetes integration. Requires RBAC.
>         backend: kubernetes
>         expanded: true
>         cpu: true
>         memory: true
>     - search:
>         provider: duckduckgo
>         target: _blank
>     - kubernetes:
>         cluster:
>           show: true
>           cpu: true
>           memory: true
>           showLabel: true
>           label: "cluster"
>         nodes:
>           show: true
>           cpu: true
>           memory: true
>           showLabel: true
>   kubernetes:
>     # change mode to 'cluster' to use RBAC service account
>     mode: cluster
>   docker:
>   settings:
> ```

Oops turns out I needed metrics-server.

insuusvenerati commented:
@jameswynn Hey i'm working on utilizing your container and helm chart but I have a small issue. I could create a separate issue for this as it may be fixable in the source but thought i'd ask here just in case. 

When using the plex widget, for example, I need a token. Ideally this would not be exposed publicly but I use FluxCD in a public repo so I don't have a choice. Instead I opted to use sops to encrypt the token in a secret. My issue is using this secret to override the configMap value is difficult because configMaps are mounted as readOnly. So using envsubst in an init container isn't working. The alternative would be to allow accepting env vars at runtime in the code. Any suggestions?

jameswynn commented:
@insuusvenerati I use the same setup. I store my credentials in a shared sops secret file, then just reference them by variable name in my configurations. For instance, here is my config for Authentik:

```yaml
- Security:
    - SSO Profile:
        icon: authentik.png
        href: https://auth.${SECRET_ROOT_DOMAIN}
        description: Authentik SSO
        app: authentik
        namespace: authentik
        widget:
          type: authentik
          url: https://auth.${SECRET_ROOT_DOMAIN}
          key: ${SECRET_HOMEPAGE_AUTHENTIK_APIKEY}
        podSelector: ""
```

jameswynn commented:
As for the metrics-server issue, I will add a specific check for its existence.

insuusvenerati commented:
> @insuusvenerati I use the same setup. I store my credentials in a shared sops secret file, then just reference them by variable name in my configurations. For instance, here is my config for Authentik:
> 
> ```yaml
> - Security:
>     - SSO Profile:
>         icon: authentik.png
>         href: https://auth.${SECRET_ROOT_DOMAIN}
>         description: Authentik SSO
>         app: authentik
>         namespace: authentik
>         widget:
>           type: authentik
>           url: https://auth.${SECRET_ROOT_DOMAIN}
>           key: ${SECRET_HOMEPAGE_AUTHENTIK_APIKEY}
>         podSelector: ""
> ```

Hmm, that's what i'm doing but the env var ends up blank in the config. Could it be because i'm using existingConfigMap instead of defining them in values.yaml?

jameswynn commented:
Unlikely, as I am doing the same. I build my configmap with kustomize and haven't had any issues. Are your other variables being subbed correctly? If your repo is public then I can take a look and see if I notice something.

Here is my kustomize config:

```yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - helm-release.yaml
configMapGenerator:
  - name: homepage
    files:
      - config/bookmarks.yaml
      - config/services.yaml
      - config/settings.yaml
      - config/widgets.yaml
      - config/docker.yaml
      - config/kubernetes.yaml
namespace: default
generatorOptions:
  disableNameSuffixHash: true
```

insuusvenerati commented:
Defining the config in values.yaml didn't work either :D Using a configMapGenerator may actually be what's letting it work. I'll give it a shot. My config lives in `cluster/apps/default/homepage` of https://github.com/insuusvenerati/k8s-home

jameswynn commented:
@clbx I tried the longhorn response you posted and it worked just fine for me. Perhaps try forcing your deployment to pull down the latest version again. If this fork stays open for very long I'll have to start versioning it.

jameswynn commented:
@insuusvenerati Check your `/cluster/flux/apps.yaml`. It is only doing variable substitutions from your top level `/cluster/config/cluster-secrets.sops.yaml`. If you define your token there then it will sub just fine.

insuusvenerati commented:
> @insuusvenerati Check your `/cluster/flux/apps.yaml`. It is only doing variable substitutions from your top level `/cluster/config/cluster-secrets.sops.yaml`. If you define your token there then it will sub just fine.

Nice catch! I'm seriously lacking experience using kustomize. I should probably fix that :D

insuusvenerati commented:
Been running on this for about 3 weeks and haven't run into any issues so far. Is there anything I can help with?

jameswynn commented:
It's been pretty stable for me as well. All I've done in the last few weeks is update the branch and resolve merge conflicts. AFAIK we are just waiting for a verdict from @benphelps.

shamoon commented:
I dont speak for Ben or the project in general. But, to me, the idea of having a large section of a codebase I can’t test myself is challenging. On the one hand maybe there are enough contributors out there and the code is stable enough that it’s a non-issue but we all know APIs and people’s ability to contribute change often.

I’m not saying I’m against merging but ya know, things to consider.

Btw, I know Ben has been otherwise occupied so I wouldn’t take the delay to mean anything more than that 👍

shamoon commented:
Ps. Just to add I think the work is great and the obviously valuable to our users. I mostly wish I knew k8s / could easily get it going for testing / debugging if / when needed.

omdv commented:
For what it's worth - I've been running it (one of previously posted containers) for several weeks now and had no issues. Waiting for PR to move to this entirely as my main k8s homepage.

jameswynn commented:
Hmmm. I could probably create some scaffolding to make testing easier. It is pretty simple to create a temp cluster using Kind (basically kubernetes inside docker). If we had that and some sample configurations then verifying it would be pretty mechanical. I haven't looked at any of the existing tests.

onedr0p commented:
@jameswynn I have a few questions on the implementation, is the current yaml still looking like this?

```yaml
ingress:
  main:
    enabled: true
    labels:
      homepage/enabled: "true"
    annotations:
      homepage/name: "Homepage"
      homepage/description: "A modern, secure, highly customizable application dashboard."
      homepage/group: "My Group"
      homepage/icon: "homepage.png"
```

If so, I am wondering why you don't move everything to annotations instead of having labels for `homepage/enabled`, also I would probably consider making the annotation key names a bit more specific.

Maybe something like this instead?

```yaml
ingress:
  main:
    enabled: true
    annotations:
      gethomepage.dev/enabled: "true"
      gethomepage.dev/name: "Homepage"
      gethomepage.dev/description: "A modern, secure, highly customizable application dashboard."
      gethomepage.dev/group: "My Group"
      gethomepage.dev/icon: "homepage.png"
```

Thanks for working on this PR and while I use [Hajimari](https://github.com/toboshii/hajimari) instead, I would give this a shot as well if merged.

jameswynn commented:
@onedr0p I've actually migrated to something nearly identical to what you've posted, though I haven't switched the "enabled" label simply because it allowed the cluster to do the filtering itself rather than in the client. It would be a rather simple change though and would be worth the effort for uniformity.

I need to write some documentation in the branch about how it works and how to get started. Much of the information in this thread is outdated at this point.

onedr0p commented:
Awesome, I actually tried PRing some changes to [Flame](https://github.com/pawelmalak/flame/pull/291) but that project looks neglected/dead. I am glad to see there's some good activity over here. If I find time I will try to test your changes.

jameswynn commented:
I just made that change from label to annotation and added some basic documentation in the `kubernetes.md` file. I also pushed an image of the new version to ghcr.io/jameswynn/homepage:kubernetes

jameswynn commented:
I found some issues with widgets in discovered services and fixed them. Published a new image.

jameswynn commented:
@shamoon what do you think about adding an abstraction on top of service discovery and statistics? If we define an API contract there then it would make it much easier to encapsulate the logic for docker and kubernetes and keep them in sync without breaking each other. Right now I'm basically reverse engineering the docker stuff to define the kubernetes pieces, so every time the docker stuff changes (which was only once, to be fair) I have to re-examine a lot of things.

shamoon commented:
> @shamoon what do you think about adding an abstraction on top of service discovery and statistics? 

Interesting but sounds like something I might also want to wait for other maintainers’ input rather than merging unilaterally. Up to you of course

jameswynn commented:
I've added some scripts that will spin up temporary kubernetes clusters on top of docker for local testing. Hopefully this will lower the maintenance bar a bit.

JazzFisch commented:
> I've added some scripts that will spin up temporary kubernetes clusters on top of docker for local testing. Hopefully this will lower the maintenance bar a bit.

Wow.  I'll look in to this soonish @jameswynn.  I'm in a lot of meetings this week, which take a lot out of me, so I'm not sure I'll be able to look until Friday at the earliest.

JazzFisch commented:
@jameswynn, I know how much work you put in to this effort, but should the helm charts be moved in to this repo as well so that the helm scripts reference something like https://homepage.github.io/helm-charts?

After running `xdg-open http://homepage.k3d.localhost:8080/` in my WSL setup I get a 404 and k9s looks like this.
![image](https://user-images.githubusercontent.com/3441425/212773020-793815ad-dcb8-4ef8-95af-3b5e93996bd6.png)

If, however, I use `tilt up` everything works and looks fine in k9s and the page loads up looking great.

![image](https://user-images.githubusercontent.com/3441425/212773639-ff9f7e9d-f92a-43b1-ac08-a923eabe2d26.png)

Do you know why k3d-deploy.sh might not be working for my WSL2 setup?

That being said, everything works great, even in my WSL2 setup!  The helper scripts in the k3d directory went a very long way to my being able to test.  The only quibble I have is that running k3d-deploy.sh shows an error because it can't find the Dockerfile when run from the k3d directory.  I think the entry should look something like this instead, yes?
`DOCKER_BUILDKIT=1 docker build -t k3d-registry.localhost:55000/homepage:local ..`

jameswynn commented:
Thanks for catching the issue with the docker build. I didn't check it after cleaning things up. The k3d-deploy.sh script is having some issues for me as well. I must have changed something after I tweaked things for Tilt.

The homepage repo would definitely be the best long term home for the helm chart, but it really depends on how much appetite the other maintainers have for it. I don't mind maintaining it in either location.

jameswynn commented:
@JazzFisch There was an issue with the dotnext PVC that prevented normal deployments for me, but that wouldn't cause your issue with pulling images. Do you have a port conflict on port 55000? Try checking `docker ps` locally to see if there are any issues with the image registry created by `k3d-up.sh`.

JazzFisch commented:
> Thanks for catching the issue with the docker build. I didn't check it after cleaning things up. The k3d-deploy.sh script is having some issues for me as well. I must have changed something after I tweaked things for Tilt.
> 
> The homepage repo would definitely be the best long term home for the helm chart, but it really depends on how much appetite the other maintainers have for it. I don't mind maintaining it in either location.

Something we can look at doing down the road.  I'm just appreciative of the work.  😄

JazzFisch commented:
> @JazzFisch There was an issue with the dotnext PVC that prevented normal deployments for me, but that wouldn't cause your issue with pulling images. Do you have a port conflict on port 55000? Try checking `docker ps` locally to see if there are any issues with the image registry created by `k3d-up.sh`.

Don't see anything obvious when I run `docker ps`:
> 12c197ce9418   registry:2                       "/entrypoint.sh /etc…"   8 minutes ago   Up 7 minutes   0.0.0.0:55000->5000/tcp

Accessing http://localhost:55000/ returned a 200, but with no content.

Running `kubectl get events` returned the following nugget:
`
9m15s       Warning   Failed                    pod/homepage-68d8fd9db6-rgvc4    Failed to pull image "k3d-registry.localhost:55000/homepage:local": rpc error: code = Unknown desc = failed to pull and unpack image "k3d-registry.localhost:55000/homepage:local": failed to resolve reference "k3d-registry.localhost:55000/homepage:local": failed to do request: Head "https://k3d-registry.localhost:55000/v2/homepage/manifests/local": dial tcp: lookup k3d-registry.localhost on 127.0.0.11:53: no such host
`

This might have something to do with running in WSL2.  This issue not withstanding, I'm not going to hold things up any longer unless I find a simple solution pretty quickly.

jameswynn commented:
That's very strange. It appears to be ignoring the registry mirror configuration which tells it to look up images from `k3d-registry.localhost:55000` at `http://k3d-registry:5000`. It might be a bug between k3d and WSL2, but  a quick search didn't turn up anything useful.

What does `docker exec -it k3d-homepage-server-0 cat /etc/rancher/k3s/registries.yam` output? It *should* be:
```yaml
auths: null
configs: null
mirrors:
  k3d-registry.localhost:55000:
    endpoint:
    - http://k3d-registry:5000
  k3d-registry:5000:
    endpoint:
    - http://k3d-registry:5000
  k3d-registry:55000:
    endpoint:
    - http://k3d-registry:5000
```

I didn't set up anything special in my /etc/hosts for this, but I am running directly in Linux so there may be a difference there.

JazzFisch commented:
> I didn't set up anything special in my /etc/hosts for this, but I am running directly in Linux so there may be a difference there.

I got the exact same thing as you showed when I ran `docker exec -it k3d-homepage-server-0 cat /etc/rancher/k3s/registries.yaml`

I don't want to hold anything up because I'm fussing about in WSL.

JazzFisch commented:
@jameswynn we're going to need you to help us update the docs in the document repo if you don't mind?  I want to make sure we capture all the cool parts of the work you did such as having the configs inside the helm chart.

shamoon commented:
Appreciate all the hard work on this one and Jazz for finally getting this over the line 👏

Aesop7 commented:
Sorry if this isn't the best place to post this, please let me know and I can move this to a separate topic.

I've been anxiously awaiting this to get merged and I'm thrilled that it did!  I started using it today and so far this has been heads and shoulders above the other things I've tried.  Also, the setup was much easier than other tools that are out there.   I did run into two issues, and sorry if I missed these in the docs:

1)  Is there a way to adjust the widget fields when using ingress annotations?

    For example, the PiHole widget allows for customization of three fields, ```["queries", "blocked", "gravity"]``` , if I wanted to only select `queries` and `blocked` using the annotation on the ingress, what's the proper syntax? I tried some variations of ```gethomepage.dev/widget.fields: "['queries','blocked']"```  but they didn't work.  

2) I'm not currently seeing the "running" or "green dot" badges for each service... should that be included as part of this feature, and if so, what is the mechanism by which it works?  (I currently just see a small outline in the ui where "running" should be, but no actual content)

Again, this feature is outstanding.  THANK YOU to everyone who helped make this a reality! 🎉

jameswynn commented:
@Aesop7 I didn't test field selection specifically, but I'll take a look into that. It's possible that additional parsing is needed. As for the "running" badge, it should show up if it is detecting the service. If the "app" and "namespace" fields are correct in your annotations then it should be able to locate the pods. Double check that the pod has a matching label "app.kubernetes.io/name=<app>".

Aesop7 commented:
Ah, okay I think I probably need to update my manifests -- seems like i'm using an older way of doing things... the manifests I have look a lot like [this](https://github.com/GoogleCloudPlatform/microservices-demo/blob/main/release/kubernetes-manifests.yaml), which as you can see only have "app" under the labels for the deployment

jameswynn commented:
@Aesop7  you can also specify your own podSelector if you need to do something different. Its in the docs here: https://gethomepage.dev/en/configs/kubernetes/

Aesop7 commented:
@jameswynn  in my case I'll just update the labels, it's relatively easy to do and I should probably be following best-practices anyway :sweat_smile:
